src/app/clusters/resource-monitoring-server/resource-monitoring-server.cpp

```diff
@@ -0,0 +1,450 @@
+    auto status = PreResetCondition();
+    if (status != Status::Success)
+    {
+        return status;
+    }
+    // Handle the reset of the condition attribute, if supported
+    if (emberAfContainsAttribute(GetEndpointId(), mClusterId, Attributes::Condition::Id))
+    {
+        if (GetDegradationDirection() == DegradationDirectionEnum::kDown)
+        {
+            UpdateCondition(100);
+        }
+        else if (GetDegradationDirection() == DegradationDirectionEnum::kUp)
+        {
+            UpdateCondition(0);
+        }
+    }
+
+    // handle the reset of the ChangeIndication attribute, mandatory
+    UpdateChangeIndication(ChangeIndicationEnum::kOk);
+
+    // Handle the reset of the LastChangedTime attribute, if supported
+    if (emberAfContainsAttribute(GetEndpointId(), mClusterId, Attributes::LastChangedTime::Id))
+    {
+        System::Clock::Milliseconds64 currentUnixTimeMS;
+        System::Clock::ClockImpl clock;
+        CHIP_ERROR err = clock.GetClock_RealTimeMS(currentUnixTimeMS);
+        if (err == CHIP_NO_ERROR)
+        {
+            System::Clock::Seconds32 currentUnixTime = std::chrono::duration_cast<System::Clock::Seconds32>(currentUnixTimeMS);
+            UpdateLastChangedTime(DataModel::MakeNullable(currentUnixTime.count()));
+        }
+    }
+
+    // call apllication specific post reset logic
+    status = PostResetCondition();
+    return status;
+}
+
+Status Instance::PreResetCondition()
+{
+    ChipLogDetail(Zcl, "ResourceMonitoringServer::PreResetCondition()");
+    return Status::Success;
+}
+
+Status Instance::PostResetCondition()
+{
+    ChipLogDetail(Zcl, "ResourceMonitoringServer::PostResetCondition()");
+    return Status::Success;
+}
+
+// This method is called by the interaction model engine when a command destined for this instance is received.
+void Instance::InvokeCommand(HandlerContext & handlerContext)
+{
+    ChipLogDetail(Zcl, "ResourceMonitoring Instance::InvokeCommand");
+    switch (handlerContext.mRequestPath.mCommandId)
+    {
+    case ResourceMonitoring::Commands::ResetCondition::Id:
+        ChipLogDetail(Zcl, "ResourceMonitoring::Commands::ResetCondition");
+
+        HandleCommand<ResourceMonitoring::Commands::ResetCondition::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleResetCondition(ctx, commandData); });
+        break;
+    }
+}
+
+// List the commands supported by this instance.
+CHIP_ERROR Instance::EnumerateAcceptedCommands(const ConcreteClusterPath & cluster,
+                                               CommandHandlerInterface::CommandIdCallback callback, void * context)
+{
+    ChipLogDetail(Zcl, "resourcemonitoring: EnumerateAcceptedCommands");
+    if (mResetConditionCommandSupported)
+    {
+        callback(ResourceMonitoring::Commands::ResetCondition::Id, context);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::ReadHepaFilterReplacableProductList(AttributeValueEncoder & aEncoder)
```

Comment:
We shouldn't have per-type functions here...  This makes anyone who implements any of these types pay to codesize cost for _all_ of them.

The right answer is that the delegate callback needs to be type-erased so that the types do not depend on the cluster id.